core/api/src/app/wallets/cancel-invoice-for-wallet.ts

```diff
@@ -0,0 +1,61 @@
+import { InvoiceAlreadyProcessedError } from "@/domain/wallet-invoices/errors"
+import { AccountValidator } from "@/domain/accounts"
+import { checkedToWalletId } from "@/domain/wallets"
+
+import {
+  AccountsRepository,
+  WalletInvoicesRepository,
+  WalletsRepository,
+} from "@/services/mongoose"
+import { LndService } from "@/services/lnd"
+import { LockService } from "@/services/lock"
+
+export const cancelInvoiceForWallet = async ({
+  walletId,
+  paymentHash,
+}: CancelInvoiceForWalletArgs): Promise<true | ApplicationError> => {
+  const checkedWalletId = checkedToWalletId(walletId)
+  if (checkedWalletId instanceof Error) return checkedWalletId
+
+  const wallet = await WalletsRepository().findById(checkedWalletId)
+  if (wallet instanceof Error) return wallet
+
+  const account = await AccountsRepository().findById(wallet.accountId)
+  if (account instanceof Error) return account
+
+  // validates account is active
+  const accountValidator = AccountValidator(account)
+  if (accountValidator instanceof Error) return accountValidator
+
+  return LockService().lockPaymentHash(paymentHash, async () => {
+    const walletInvoices = WalletInvoicesRepository()
+    const walletInvoice = await walletInvoices.findForWalletByPaymentHash({
+      walletId,
+      paymentHash,
+    })
+    if (walletInvoice instanceof Error) return walletInvoice
+    if (walletInvoice.processingCompleted) {
+      return new InvoiceAlreadyProcessedError()
+    }
+
+    const { pubkey } = walletInvoice
+    const lndService = LndService()
+    if (lndService instanceof Error) return lndService
+
+    const lnInvoiceLookup = await lndService.lookupInvoice({ pubkey, paymentHash })
+    if (lnInvoiceLookup instanceof Error) return lnInvoiceLookup
+
+    const { isSettled, isHeld } = lnInvoiceLookup
+    if (isSettled || isHeld) {
```

Comment:
should we allow to cancel held invoices too? cc @nicolasburtey

Reply:
held invoices are just because it was not processed by our system yet? (because of a trigger restart for instance)

in that case I don't think so. it would expose some of our internal system